contracts/exchangeIssuance/FlashMintDex.sol

```diff
@@ -0,0 +1,612 @@
+     *
+     * @return totalEthNeeded  Amount of input tokens required to perfrom the issuance
+     */
+    function getIssueExactSet(
+        IssueRedeemParams memory _issueParams,
+        DEXAdapterV2.SwapData memory _swapDataInputTokenToWeth
+    )
+        external
+        returns (uint256)
+    {
+        uint256 totalEthNeeded = 0;
+        (,, uint256[] memory wethCosts) = _getWethCostsPerComponent(_issueParams);
+
+        for (uint256 i = 0; i < wethCosts.length; i++) {
+            totalEthNeeded += wethCosts[i];
+        }
+        return dexAdapter.getAmountIn(_swapDataInputTokenToWeth, totalEthNeeded);
+    }
+
+    /**
+     * Gets the amount of specified payment token expected to be received after redeeming 
+     * a given quantity of set token with the provided redemption params.
+     * This function is not marked view, but should be static called from frontends.
+     * This constraint is due to the need to interact with the Uniswap V3 quoter contract
+     *
+     * @param _redeemParams                Struct containing addresses, amounts, and swap data for redemption
+     * @param _swapDataWethToOutputToken   Swap data to trade WETH for output token. Use empty swap data if output token is ETH or WETH. 
+     *
+     * @return                             the amount of output tokens expected after performing redemption
+     */
+    function getRedeemExactSet(
+        IssueRedeemParams memory _redeemParams,
+        DEXAdapterV2.SwapData memory _swapDataWethToOutputToken
+    )
+        external
+        returns (uint256)
+    {
+        uint256 totalWethReceived = 0;
+        (,, uint256[] memory wethReceived) = _getWethReceivedPerComponent(_redeemParams);
+
+        for (uint256 i = 0; i < wethReceived.length; i++) {
+            totalWethReceived += wethReceived[i];
+        }
+        return dexAdapter.getAmountOut(_swapDataWethToOutputToken, totalWethReceived);
+    }
+
+    /**
+    * Issues an exact amount of SetTokens for given amount of ETH.
+    * Any leftover ether is returned to the caller.
+    *
+    * @param _issueParams           Struct containing addresses, amounts, and swap data for issuance
+    *
+    * @return ethSpent              Amount of ETH spent
+    */
+    function issueExactSetFromETH(IssueRedeemParams memory _issueParams)
+        external
+        payable
+        isValidModuleAndSet(_issueParams.issuanceModule, address(_issueParams.setToken))
+        nonReentrant
+        returns (uint256 ethSpent)
+    {
+        require(msg.value > 0, "FlashMint: NO ETH SENT");
+
+        IWETH(WETH).deposit{value: msg.value}();
+
+        uint256 ethUsedForIssuance = _issueExactSetFromWeth(_issueParams);
+
+        require(ethUsedForIssuance <= msg.value, "FlashMint: OVERSPENT ETH");
+
+        uint256 ethReturned = msg.value.sub(ethUsedForIssuance);
+        if (ethReturned > 0) {
```

Comment:
In practice there will always be leftover ETH, so the check here might be wasted gas.